packages/tailwindcss/src/ast.test.ts

```diff
@@ -95,3 +97,88 @@ it('should stop walking when returning `WalkAction.Stop`', () => {
     }
   `)
 })
+
+it('should not emit empty rules once optimized', () => {
+  let ast = CSS.parse(css`
+    /* Empty rule */
+    .foo {
+    }
+
+    /* Empty rule, with nesting */
+    .foo {
+      .bar {
+      }
+      .baz {
+      }
+    }
+
+    /* Empty rule, with special case '&' rules */
+    .foo {
+      & {
+        &:hover {
+        }
+        &:focus {
+        }
+      }
+    }
+
+    /* Empty at-rule */
+    @media (min-width: 768px) {
+    }
+
+    /* Empty at-rule with nesting*/
+    @media (min-width: 768px) {
+      .foo {
+      }
+
+      @media (min-width: 1024px) {
+        .bar {
+        }
+      }
+    }
+
+    /* Exceptions: */
+    @charset 'UTF-8';
+    @layer foo, bar, baz;
+    @custom-media --modern (color), (hover);
+  `)
+
+  expect(toCss(ast)).toMatchInlineSnapshot(`
+    ".foo {
+    }
+    .foo {
+      .bar {
+      }
+      .baz {
+      }
+    }
+    .foo {
+      & {
+        &:hover {
+        }
+        &:focus {
+        }
+      }
+    }
+    @media (min-width: 768px);
+    @media (min-width: 768px) {
+      .foo {
+      }
+      @media (min-width: 1024px) {
+        .bar {
+        }
+      }
+    }
+    @charset 'UTF-8';
+    @layer foo, bar, baz;
+    @custom-media --modern (color), (hover);
+    "
+  `)
+
+  expect(toCss(optimizeAst(ast))).toMatchInlineSnapshot(`
+    "@charset 'UTF-8';
+    @layer foo, bar, baz;
+    @custom-media --modern (color), (hover);
```

Comment:
```suggestion
    @custom-media --modern (color), (hover);
    @namespace "http://www.w3.org/1999/xhtml";
```

Reply:
(I think I got the syntax right)

Reply:
(you did not, but it's fine 😂)

Reply:
Or you know what, it looks like `"` is required for `@charset` and can't have `'`. MDN doesn't say anything about `"` vs `'` for `@namespace`

Reply:
Note: we just parse the value as-is, so if you use `@charset 'foo'` we keep it as-is.